DESIGN.md

```diff
@@ -0,0 +1,98 @@
+# Bulk Data Extract (BDE) Processor design
+
+## User requirements
+
+-  The _user interface_ is a read-only PostgreSQL database in AWS RDS, with
+   Azure authentication.
+-  _Accessible_ to all
+   [current BDE users](https://github.com/linz/ds-infra-team/wiki/BDE-users).
+-  Process BDE files
+   _[every day](https://toitutewhenua.atlassian.net/wiki/spaces/GEOD/pages/86418416/BDE+Processor+Product)
+   except Sunday_ into a structure conforming to the legacy BDE Processor DB.
+   The processing should take _much_ less than a day.
+-  Optional _post-processing_ pulled in from separate internal repository.
+-  All changes for a single day are imported within a _single transaction._
+-  _Immediate notification_ of completed import (like the legacy BDE Processor
+   SNS topic).
+-  Schema changes synced with users.
+
+## Technical requirements
+
+-  Keep BDE files. This could be done as a separate system, but that would
+   complicate the architecture and release process. **TBD:** How long do we need
+   to keep these files? Indefinitely is going to be very costly. A year would be
+   cheaper, but completely arbitrary. Keeping just the last L0 and the
+   associated L5 files would require some code, but would probably be the
+   minimum we could keep to enable a restore from a blank slate.
+-  Thorough testing of every part of the process before merging any pull
+   requests.
+-  Separate, managed accounts for infrastructure testing and production.
+-  Trivial deployment.
+-  Monitor performance characteristics.
+-  Verbose logging to central ElasticSearch instance.
+-  Alerts:
+   -  Process start/finish with result and auxiliary information in case of
+      failure.
+   -  Big subprocesses' start/finish.
+
+## Wanted features
+
+-  _Pull rather than push data from Landonline._ This way we can keep the entire
+   system together, rather than having a costly and awkward split between
+   on-premise and cloud architecture.
+-  _Continuous deployment._ Any time a change is merged to the main branch some
+   smoke tests are performed and the result is deployed to production.
+-  _React to change rather than run on a timer,_ where possible. Ideally the BDE
+   Processor should know about new files as soon as the last file has been
+   closed for writing, and should immediately start processing.
+-  _From-scratch deployment friendly._ That is, we should always have the option
+   to deploy from scratch, deleting _all_ the resources we've created and
+   getting back to an identical state (from the end users' perspective) in a
+   timely fashion (that is, before the next data dump).
+-  _Access for everybody at the org._ This should be simpler than the current
+   solution, and should be easy with read replicas.
+-  _Public access,_ which in turn requires:
```

Comment:
What's the difference? They would simply be allowed to run read-only queries against the CC-licensed data in our DB.

Reply:
It would be worth understanding the typical user profile and use case for this and thinking about how much this could cost us (if we need lots of load balancing / scalability etc), but we could also look at other orgs doing similar things to get that info, like [this](https://docs.rfam.org/en/latest/database.html).

This just came out a discussion we had the other day when looking at the various tricking options for providing read access for LINZ internal people, about whether we could just make it public.

Reply:
I'm not sure I understand. Public access would include access to councils and other customers, so it would solve several problems in one go. Of course, we'd have to determine how to do this responsibly.